core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go

```diff
@@ -39,52 +41,74 @@ type EVMAutomationUpkeepResult21 struct {
 	Retryable        bool
 }
 
-// TODO: align once we merge with ocr2keepers new types (ocr2keepers.CheckResult)
-func (enc EVMAutomationEncoder21) EncodeReport(toReport []ocr2keepers.UpkeepResult) ([]byte, error) {
-	if len(toReport) == 0 {
-		return nil, nil
+type EVMAutomationResultExtension21 struct {
+	FastGasWei    *big.Int
+	LinkNative    *big.Int
+	FailureReason uint8 // this is not encoded, only pass along for the purpose of pipeline run
+}
+
+func (enc EVMAutomationEncoder21) Encode(results ...ocr2keepers.CheckResult) ([]byte, error) {
+	if len(results) == 0 {
+		return nil, ErrEmptyResults
 	}
 
 	report := automation_utils_2_1.KeeperRegistryBase21Report{
-		UpkeepIds:    make([]*big.Int, len(toReport)),
-		GasLimits:    make([]*big.Int, len(toReport)),
-		Triggers:     make([][]byte, len(toReport)),
-		PerformDatas: make([][]byte, len(toReport)),
+		UpkeepIds:    make([]*big.Int, len(results)),
+		GasLimits:    make([]*big.Int, len(results)),
+		Triggers:     make([][]byte, len(results)),
+		PerformDatas: make([][]byte, len(results)),
 	}
 
-	for i, result := range toReport {
-		res, ok := result.(EVMAutomationUpkeepResult21)
+	for i, result := range results {
+		ext, ok := result.Extension.(EVMAutomationResultExtension21)
 		if !ok {
-			return nil, fmt.Errorf("unexpected upkeep result struct")
+			return nil, fmt.Errorf("unexpected check result extension struct")
 		}
 
 		// only take these values from the first result
 		// TODO: find a new way to get these values
 		if i == 0 {
-			report.FastGasWei = res.FastGasWei
-			report.LinkNative = res.LinkNative
+			report.FastGasWei = ext.FastGasWei
```

Comment:
not related to this PR but can we create a ticket for the above TODO? We should take the fastGas and linkNative from the result having the highest checkBlockNumber